pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1341,108 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
+		if err != nil {
+			if !apiErrorPkg.IsNotFound(err) {
+				return "", err
+			}
+		}
+	}
+
+	// Check if global pull secret from env as it comes from hive config
+	globalPullSecret := os.Getenv("GLOBAL_PULL_SECRET")
+
+	switch {
+	case globalPullSecret != "" && localPullSecret != "":
+		// Merge local pullSecret and globalPullSecret. If both pull secrets have same registry name
+		// then the merged pull secret will have registry secret from local pull secret
+		pullSecret, err := controllerutils.MergeJsons(globalPullSecret, localPullSecret, cdLog)
+		if err != nil {
+			errMsg := "unable to merge global pull secret with local pull secret"
+			cdLog.WithError(err).Error(errMsg)
+			return "", errors.Wrap(err, errMsg)
+		}
+		return pullSecret, nil
+	case globalPullSecret != "":
+		return globalPullSecret, nil
+	case localPullSecret != "":
+		return localPullSecret, nil
+	default:
+		errMsg := "clusterdeployment must specify pull secret since hiveconfig does not specify a global pull secret"
+		cdLog.Error(errMsg)
+		return "", errors.New(errMsg)
+	}
+}
+
+// updatePullSecretInfo adds pull secret information in cluster deployment and cluster deployment status
+func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+	var existingSecretName, existingPullSecret string
+	var err error
+
+	// Get existing pull secret
+	pullSecretObjExists := true
+	existingPullSecretObj := &corev1.Secret{}
+	if cd.Status.PullSecret.Name != "" {
+		existingSecretName = cd.Status.PullSecret.Name
+		existingPullSecret, err = controllerutils.LoadSecretData(r.Client, existingSecretName, cd.Namespace, corev1.DockerConfigJsonKey)
+		if err != nil {
+			pullSecretObjExists = false
+		}
+	} else {
+		pullSecretObjExists = false
+	}
+
+	if pullSecretObjExists {
+		if controllerutils.GetHashOfPullSecret(existingPullSecret) == controllerutils.GetHashOfPullSecret(pullSecret) {
+			cdLog.Debug("No new pull secret found")
+			return nil
+		}
+	}
+
+	// create a new pull secret object and update the cluster deployment status
+	newPullSecretObj := generatePullSecretObj(
+		pullSecret,
+		apiHelpers.GetResourceName(cd.Name, "merged-"+utilrand.String(8)+"-"),
+		cd,
+	)
+	newSecretName := newPullSecretObj.GetGenerateName()
+	err = r.Create(context.TODO(), newPullSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "error creating new pull secret object")
+	}
+	defer func() {
+		if err != nil {
+			if err := r.Delete(context.TODO(), newPullSecretObj); err != nil {
+				cdLog.WithError(err).Errorf("Error deleting the new pull secret %s", newSecretName)
+			}
+		}
+	}()
+	cdLog.WithField("secretName", newSecretName).Info("Created new merged pull secret for cluster deployment")
+
+	// Update new pullsecret information in cluster deployment status
+	cd.Status.PullSecret = corev1.LocalObjectReference{Name: newSecretName}
```

Comment:
The name referenced should be `newPullSecretObj.Name` not `newPullSecretObj.GetGenerateName()`.